File: androidApp/src/androidTest/java/com/mbta/tid/mbta_app/android/component/SaveFavoritesFlowTest.kt
package com.mbta.tid.mbta_app.android.component

import androidx.compose.ui.test.ExperimentalTestApi
import androidx.compose.ui.test.assertIsDisplayed
import androidx.compose.ui.test.assertIsNotEnabled
import androidx.compose.ui.test.hasText
import androidx.compose.ui.test.junit4.createComposeRule
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import com.mbta.tid.mbta_app.android.loadKoinMocks
import com.mbta.tid.mbta_app.android.testUtils.waitUntilDefaultTimeout
import com.mbta.tid.mbta_app.android.testUtils.waitUntilExactlyOneExistsDefaultTimeout
import com.mbta.tid.mbta_app.model.Direction
import com.mbta.tid.mbta_app.model.FavoriteSettings
import com.mbta.tid.mbta_app.model.LineOrRoute
import com.mbta.tid.mbta_app.model.RouteStopDirection
import com.mbta.tid.mbta_app.usecases.EditFavoritesContext
import com.mbta.tid.mbta_app.utils.TestData
import com.mbta.tid.mbta_app.viewModel.IToastViewModel
import com.mbta.tid.mbta_app.viewModel.MockToastViewModel
import com.mbta.tid.mbta_app.viewModel.ToastViewModel
import dev.mokkery.MockMode
import dev.mokkery.mock
import dev.mokkery.verify
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import kotlinx.coroutines.runBlocking
import org.junit.Rule
import org.junit.Test

@OptIn(ExperimentalTestApi::class)
class SaveFavoritesFlowTest {

    @get:Rule val composeTestRule = createComposeRule()

    val line =
        LineOrRoute.Line(
            TestData.getLine("line-Green"),
            setOf(
                TestData.getRoute("Green-B"),
                TestData.getRoute("Green-C"),
                TestData.getRoute("Green-D"),
                TestData.getRoute("Green-E"),
            ),
        )
    val stop = TestData.getStop("place-boyls")
    val direction0 = Direction(id = 0, name = "West", destination = "Copley & West")
    val direction1 = Direction(id = 1, name = "East", destination = "Park St & North")
    val directions = listOf(direction0, direction1)

    @Test
    fun testWithoutTappingAnyButtonSavesProposedChanges() {
        var updateFavoritesCalledFor: Map<RouteStopDirection, FavoriteSettings?> = mapOf()
        var onCloseCalled = false

        composeTestRule.setContent {
            FavoriteConfirmationDialog(
                lineOrRoute = line,
                stop = stop,
                selectedDirection = 0,
                directions = directions,
                proposedFavorites = mapOf(0 to FavoriteSettings()),
                context = EditFavoritesContext.Favorites,
                updateFavorites = { updateFavoritesCalledFor = (it) },
            ) {
                onCloseCalled = true
            }
        }

        composeTestRule.onNodeWithText("Add").performClick()
        composeTestRule.waitForIdle()
        assertEquals(
            updateFavoritesCalledFor,
            mapOf(RouteStopDirection(line.id, stop.id, 0) to FavoriteSettings()),
        )
        assertTrue(onCloseCalled)
    }

    @Test
    fun testCancelDoesntUpdateFavorites() {
        var updateFavoritesCalled = false
        var onCloseCalled = false

        composeTestRule.setContent {
            FavoriteConfirmationDialog(
                lineOrRoute = line,
                stop = stop,
                directions = directions,
                selectedDirection = 0,
                proposedFavorites = mapOf(0 to FavoriteSettings()),
                context = EditFavoritesContext.Favorites,
                updateFavorites = { updateFavoritesCalled = true },
            ) {
                onCloseCalled = true
            }
        }

        composeTestRule.onNodeWithText("Cancel").performClick()
        composeTestRule.waitForIdle()
        assertTrue(onCloseCalled)
        assertFalse(updateFavoritesCalled)
    }

    @Test
    fun testAddingOtherDirectionSavesBoth() {
        var updateFavoritesCalledFor: Map<RouteStopDirection, FavoriteSettings?> = mapOf()

        composeTestRule.setContent {
            FavoriteConfirmationDialog(
                lineOrRoute = line,
                stop = stop,
                directions = directions,
                selectedDirection = 0,
                proposedFavorites = mapOf(0 to FavoriteSettings()),
                context = EditFavoritesContext.Favorites,
                updateFavorites = { updateFavoritesCalledFor = it },
            ) {}
        }

        composeTestRule.onNodeWithText("East", substring = true).performClick()
        composeTestRule.onNodeWithText("Add").performClick()
        composeTestRule.waitForIdle()
        assertEquals(
            updateFavoritesCalledFor,
            mapOf(
                RouteStopDirection(line.id, stop.id, 0) to FavoriteSettings(),
                RouteStopDirection(line.id, stop.id, 1) to FavoriteSettings(),
            ),
        )
    }

    @Test
    fun testRemovingOtherDirectoinSavesBoth() {
        var updateFavoritesCalledFor: Map<RouteStopDirection, FavoriteSettings?> = mapOf()

        composeTestRule.setContent {
            FavoriteConfirmationDialog(
                lineOrRoute = line,
                stop = stop,
                directions = directions,
                selectedDirection = 0,
                proposedFavorites = mapOf(0 to FavoriteSettings(), 1 to FavoriteSettings()),
                context = EditFavoritesContext.Favorites,
                updateFavorites = { updateFavoritesCalledFor = it },
            ) {}
        }

        composeTestRule.onNodeWithText("East", substring = true).performClick()
        composeTestRule.onNodeWithText("Add").performClick()
        composeTestRule.waitForIdle()
        assertEquals(
            updateFavoritesCalledFor,
            mapOf(
                RouteStopDirection(line.id, stop.id, 0) to FavoriteSettings(),
                RouteStopDirection(line.id, stop.id, 1) to null,
            ),
        )
    }

    @Test
    fun testRemovingProposedFavoriteDisablesAddButton() {
        var updateFavoritesCalledFor: Map<RouteStopDirection, FavoriteSettings?> = mapOf()
        var onCloseCalled = false

        composeTestRule.setContent {
            FavoriteConfirmationDialog(
                lineOrRoute = line,
                stop = stop,
                directions = directions,
                selectedDirection = 0,
                proposedFavorites = mapOf(0 to FavoriteSettings()),
                context = EditFavoritesContext.Favorites,
                updateFavorites = { updateFavoritesCalledFor = it },
            ) {
                onCloseCalled = true
            }
        }

        composeTestRule.onNodeWithText("West", substring = true).performClick()
        composeTestRule.onNodeWithText("Add").assertIsNotEnabled()
    }

    @Test
    fun testFavoritingOnlyDirectionPresentsDialogWhenNonBus() {
        var updateFavoritesCalledFor: Map<RouteStopDirection, FavoriteSettings?> = mapOf()
        var onCloseCalled = false

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = line,
                stop = stop,
                directions = listOf(direction0),
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                isFavorite = { false },
                updateFavorites = { updateFavoritesCalledFor = it },
                onClose = { onCloseCalled = true },
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule.waitUntilExactlyOneExistsDefaultTimeout(hasText("Add"))
        composeTestRule.onNodeWithText("Add").assertExists()
    }

    @Test
    fun testFavoritingOnlyDirectionSkipsDialogWhenBus() {
        var updateFavoritesCalledFor: Map<RouteStopDirection, FavoriteSettings?> = mapOf()
        var onCloseCalled = false

        val busRoute = TestData.getRoute("15")
        val busStop = TestData.getStop("17861")

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = LineOrRoute.Route(busRoute),
                stop = busStop,
                directions = listOf(direction0),
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                isFavorite = { false },
                updateFavorites = { updateFavoritesCalledFor = it },
                onClose = { onCloseCalled = true },
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule.waitUntilDefaultTimeout { onCloseCalled }
        assertTrue(onCloseCalled)
        assertEquals(
            updateFavoritesCalledFor,
            mapOf(RouteStopDirection(busRoute.id, busStop.id, 0) to FavoriteSettings()),
        )
    }

    @Test
    fun testUnfavoritingOnlyDirectionUpdatesFavoritesWithoutDialog() {
        var updateFavoritesCalledFor: Map<RouteStopDirection, FavoriteSettings?> = mapOf()
        var onCloseCalled = false

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = line,
                stop = stop,
                directions = listOf(direction0),
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                isFavorite = { true },
                updateFavorites = { updateFavoritesCalledFor = it },
                onClose = { onCloseCalled = true },
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule.waitUntilDefaultTimeout { onCloseCalled }
        assertTrue(onCloseCalled)
        assertEquals(
            updateFavoritesCalledFor,
            mapOf(RouteStopDirection(line.id, stop.id, 0) to null),
        )
    }

    @Test
    fun testFavoritingWhenOnlyDirectionIsOppositePresentsDialog() {
        var updateFavoritesCalledFor: Map<RouteStopDirection, FavoriteSettings?> = mapOf()
        var onCloseCalled = false

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = line,
                stop = stop,
                directions = listOf(direction1),
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                isFavorite = { false },
                updateFavorites = { updateFavoritesCalledFor = it },
                onClose = { onCloseCalled = true },
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule.onNodeWithText("Eastbound service only").assertIsDisplayed()
        composeTestRule.onNodeWithText("Add").performClick()
        composeTestRule.waitForIdle()
        composeTestRule.waitUntilDefaultTimeout { onCloseCalled }
        assertTrue(onCloseCalled)
        assertEquals(
            updateFavoritesCalledFor,
            mapOf(RouteStopDirection(line.id, stop.id, 1) to FavoriteSettings()),
        )
    }

    @Test
    fun testFavoritingSingleDirectionDisplaysToast() {
        val toastVM = MockToastViewModel()
        var displayedToast: ToastViewModel.Toast? = null
        toastVM.onShowToast = { displayedToast = it }

        loadKoinMocks(TestData)

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = line,
                stop = stop,
                directions = listOf(direction0, direction1),
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                toastViewModel = toastVM,
                isFavorite = { false },
                updateFavorites = {},
                onClose = {},
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule.waitUntilExactlyOneExistsDefaultTimeout(hasText("Add"))
        composeTestRule.onNodeWithText("Add").performClick()

        composeTestRule.waitUntilDefaultTimeout {
            displayedToast?.message ==
                "<b>Westbound Green Line</b> at <b>Boylston</b> added to Favorites"
        }
    }

    @Test
    fun testFavoritingBothDirectionsDisplaysToast() {
        val toastVM = MockToastViewModel()
        var displayedToast: ToastViewModel.Toast? = null
        toastVM.onShowToast = { displayedToast = it }

        loadKoinMocks(TestData)

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = line,
                stop = stop,
                directions = listOf(direction0, direction1),
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                toastViewModel = toastVM,
                isFavorite = { false },
                updateFavorites = {},
                onClose = {},
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule.onNodeWithText("East", substring = true).performClick()
        composeTestRule.waitUntilExactlyOneExistsDefaultTimeout(hasText("Add"))
        composeTestRule.onNodeWithText("Add").performClick()

        composeTestRule.waitUntilDefaultTimeout {
            displayedToast?.message == "<b>Green Line</b> at <b>Boylston</b> added to Favorites"
        }
    }

    @Test
    fun testFavoritingDisplaysToastWhenDialogIsSkipped() {
        val toastVM = MockToastViewModel()
        var displayedToast: ToastViewModel.Toast? = null
        toastVM.onShowToast = { displayedToast = it }

        val busRoute = TestData.getRoute("15")
        val busStop = TestData.getStop("17861")

        loadKoinMocks(TestData)

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = LineOrRoute.Route(busRoute),
                stop = busStop,
                directions = listOf(direction0),
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                toastViewModel = toastVM,
                isFavorite = { false },
                updateFavorites = {},
                onClose = {},
            )
        }

        composeTestRule.waitUntilDefaultTimeout {
            displayedToast?.message == "<b>Outbound 15 bus</b> at <b>Ruggles</b> added to Favorites"
        }
    }

    @Test
    fun testFavoritingToastFallbackText() = runBlocking {
        val toastVM = mock<IToastViewModel>(MockMode.autofill)

        val busRoute = TestData.getRoute("15")
        val busStop = TestData.getStop("17861")
        // no TestData for getting the appropriate labels for this favorite
        loadKoinMocks()

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = LineOrRoute.Route(busRoute),
                stop = busStop,
                directions = listOf(direction0),
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                toastViewModel = toastVM,
                isFavorite = { false },
                updateFavorites = {},
                onClose = {},
            )
        }
        composeTestRule.awaitIdle()

        verify {
            toastVM.showToast(
                ToastViewModel.Toast("Added to Favorites", duration = ToastViewModel.Duration.Short)
            )
        }
    }

    @Test
    fun testFavoritingWhenDropOffOnlyPresentsDialog() {
        var onCloseCalled = false

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = line,
                stop = stop,
                directions = listOf(),
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                isFavorite = { false },
                updateFavorites = {},
                onClose = { onCloseCalled = true },
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule.onNodeWithText("This stop is drop-off only").assertIsDisplayed()
        composeTestRule.onNodeWithText("Add").assertDoesNotExist()
        composeTestRule.onNodeWithText("Okay").performClick()
        composeTestRule.waitForIdle()
        composeTestRule.waitUntilDefaultTimeout { onCloseCalled }
        assertTrue(onCloseCalled)
    }

    @Test
    fun testDialogTitleFavoritesContext() {
        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = line,
                stop = stop,
                directions = listOf(direction1),
                selectedDirection = 1,
                context = EditFavoritesContext.Favorites,
                isFavorite = { false },
                updateFavorites = {},
                onClose = {},
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule.onNodeWithText("Add Green Line at Boylston").assertIsDisplayed()
    }

    @Test
    fun testDialogTitleStopDetailsContext() {
        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = line,
                stop = stop,
                directions = listOf(direction1),
                selectedDirection = 1,
                context = EditFavoritesContext.StopDetails,
                isFavorite = { false },
                updateFavorites = {},
                onClose = {},
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule
            .onNodeWithText("Add Green Line at Boylston to Favorites")
            .assertIsDisplayed()
    }

    @Test
    fun testFavoritingWhenTwoDirectionsPresentsDialog() {
        var updateFavoritesCalledFor: Map<RouteStopDirection, FavoriteSettings?> = mapOf()
        var onCloseCalled = false

        composeTestRule.setContent {
            SaveFavoritesFlow(
                lineOrRoute = line,
                stop = stop,
                directions = directions,
                selectedDirection = 0,
                context = EditFavoritesContext.Favorites,
                isFavorite = { rsd -> rsd.direction == 1 },
                updateFavorites = { updateFavoritesCalledFor = it },
                onClose = { onCloseCalled = true },
            )
        }
        composeTestRule.waitForIdle()
        composeTestRule.onNodeWithText("Add").performClick()
        composeTestRule.waitForIdle()
        composeTestRule.waitUntilDefaultTimeout { onCloseCalled }
        assertTrue(onCloseCalled)
        assertEquals(
            updateFavoritesCalledFor,
            mapOf(
                RouteStopDirection(line.id, stop.id, 0) to FavoriteSettings(),
                RouteStopDirection(line.id, stop.id, 1) to FavoriteSettings(),
            ),
        )
    }
}

File: iosApp/iosAppTests/Views/SaveFavoritesFlowTest.swift
//
//  SaveFavoritesFlowTest.swift
//  iosAppTests
//
//  Created by Kayla Brady on 6/30/25.
//  Copyright © 2025 MBTA. All rights reserved.
//

@testable import iosApp

import Combine
import Shared
import SwiftUI
import ViewInspector
import XCTest

final class SaveFavoritesFlowTest: XCTestCase {
    override func setUp() {
        executionTimeAllowance = 60
    }

    let line = LineOrRoute.line(
        TestData.getLine(id: "line-Green"),
        Set([
            TestData.getRoute(id: "Green-B"),
            TestData.getRoute(id: "Green-C"),
            TestData.getRoute(id: "Green-D"),
            TestData.getRoute(id: "Green-E"),
        ])
    )

    let stop = TestData.getStop(id: "place-boyls")
    let direction0 = Direction(name: "West", destination: "Copley & West", id: 0)
    let direction1 = Direction(name: "East", destination: "Park St & North", id: 1)
    var directions: [Direction] { [direction0, direction1] }

    func testWithoutTappingAnyButtonSavesProposedChanges() throws {
        var updateFavoritesCalledFor: [RouteStopDirection: FavoriteSettings?] = [:]
        var onCloseCalled = false

        let sut = FavoriteConfirmationDialogActions(
            lineOrRoute: line,
            stop: stop,
            favoritesToSave: [direction0: .init()],
            updateFavorites: { updateFavoritesCalledFor = $0 },
            onClose: { onCloseCalled = true },
        )

        ViewHosting.host(view: sut.withFixedSettings([:]))

        try sut.inspect().find(button: "Add").tap()

        XCTAssertEqual(updateFavoritesCalledFor, [.init(route: line.id, stop: stop.id, direction: 0): .init()])
        XCTAssertTrue(onCloseCalled)
    }

    func testCancelDoesntUpdateFavorites() throws {
        var updateFavoritesCalled = false
        var onCloseCalled = false

        let sut = FavoriteConfirmationDialogActions(
            lineOrRoute: line,
            stop: stop,
            favoritesToSave: [direction0: .init()],
            updateFavorites: { _ in updateFavoritesCalled = true },
            onClose: { onCloseCalled = true },
        )

        ViewHosting.host(view: sut.withFixedSettings([:]))

        try sut.inspect().find(button: "Cancel").tap()

        XCTAssertTrue(onCloseCalled)
        XCTAssertFalse(updateFavoritesCalled)
    }

    func testAddingOtherDirectionUpdates() throws {
        var updateLocalFavoriteCalledFor: [Direction: FavoriteSettings?] = [:]

        let sut = FavoriteConfirmationDialogContents(
            lineOrRoute: line,
            stop: stop,
            directions: directions,
            selectedDirection: 0,
            context: EditFavoritesContext.favorites,
            favoritesToSave: [direction0: .init(), direction1: nil],
            updateLocalFavorite: { updateLocalFavoriteCalledFor = [$0: $1] },
        )

        ViewHosting.host(view: sut.withFixedSettings([:]))

        try sut.inspect().findAll(ViewType.Button.self)[1].tap()

        XCTAssertEqual(updateLocalFavoriteCalledFor, [direction1: .init()])
    }

    func testRemovingOtherDirectionUpdates() throws {
        var updateLocalFavoriteCalledFor: [Direction: FavoriteSettings?] = [:]

        let sut = FavoriteConfirmationDialogContents(
            lineOrRoute: line,
            stop: stop,
            directions: directions,
            selectedDirection: 0,
            context: EditFavoritesContext.favorites,
            favoritesToSave: [direction0: .init(), direction1: .init()],
            updateLocalFavorite: { updateLocalFavoriteCalledFor = [$0: $1] },
        )

        ViewHosting.host(view: sut.withFixedSettings([:]))

        try sut.inspect().findAll(ViewType.Button.self)[1].tap()

        XCTAssertEqual(updateLocalFavoriteCalledFor, [direction1: nil])
    }

    func testRemovingProposedFavoriteDisablesAddButton() throws {
        var updateFavoritesCalled = false
        var onCloseCalled = false

        let sut = FavoriteConfirmationDialogActions(
            lineOrRoute: line,
            stop: stop,
            favoritesToSave: [direction0: nil, direction1: nil],
            updateFavorites: { _ in updateFavoritesCalled = true },
            onClose: { onCloseCalled = true },
        )

        ViewHosting.host(view: sut.withFixedSettings([:]))

        XCTAssertTrue(try sut.inspect().find(button: "Add").isDisabled())
    }

    @MainActor
    func testFavoritingOnlyDirectionPresentsDialogWhenNonBus() throws {
        var updateFavoritesCalledFor: [RouteStopDirection: FavoriteSettings?] = [:]

        let sut = SaveFavoritesFlow(
            lineOrRoute: line,
            stop: stop,
            directions: [direction0],
            selectedDirection: 0,
            context: EditFavoritesContext.favorites,
            global: .init(objects: .init()),
            isFavorite: { _ in false },
            updateFavorites: { updateFavoritesCalledFor = $0 },
            onClose: {},
            pushNavEntry: { _ in },
        )

        let exp = sut.inspection.inspect(after: 0.5) { view in
            XCTAssertNotNil(try view.find(FavoriteConfirmationDialog.self))
        }

        ViewHosting.host(view: sut.withFixedSettings([:]))

        wait(for: [exp], timeout: 2)
    }

    func testFavoritingOnlyDirectionSkipsDialogWhenBus() throws {
        var updateFavoritesCalledFor: [RouteStopDirection: FavoriteSettings?] = [:]
        let onCloseExp = XCTestExpectation(description: "On close called")

        let route = ObjectCollectionBuilder().route { route in
            route.type = RouteType.bus
        }

        let sut = SaveFavoritesFlow(
            lineOrRoute: LineOrRoute.route(route),
            stop: stop,
            directions: [direction0],
            selectedDirection: 0,
            context: EditFavoritesContext.favorites,
            global: .init(objects: .init()),
            isFavorite: { _ in false },
            updateFavorites: { updateFavoritesCalledFor = $0 },
            onClose: { onCloseExp.fulfill() },
            pushNavEntry: { _ in },
        )

        ViewHosting.host(view: sut.withFixedSettings([:]))

        wait(for: [onCloseExp], timeout: 2)
        XCTAssertEqual(updateFavoritesCalledFor, [.init(route: route.id, stop: stop.id, direction: 0): .init()])
    }

    func testUnfavoritingOnlyDirectionUpdatesFavoritesWithoutDialog() throws {
        var updateFavoritesCalledFor: [RouteStopDirection: FavoriteSettings?] = [:]
        let onCloseExp = XCTestExpectation(description: "On close called")

        let route = ObjectCollectionBuilder().route { route in
            route.type = RouteType.bus
        }

        let sut = SaveFavoritesFlow(
            lineOrRoute: LineOrRoute.route(route),
            stop: stop,
            directions: [direction0],
            selectedDirection: 0,
            context: EditFavoritesContext.favorites,
            global: .init(objects: .init()),
            isFavorite: { _ in true },
            updateFavorites: { updateFavoritesCalledFor = $0 },
            onClose: { onCloseExp.fulfill() },
            pushNavEntry: { _ in },
        )

        ViewHosting.host(view: sut.withFixedSettings([:]))

        wait(for: [onCloseExp], timeout: 2)
        XCTAssertEqual(updateFavoritesCalledFor, [.init(route: route.id, stop: stop.id, direction: 0): nil])
    }

    @MainActor
    func testFavoritingWhenOnlyDirectionIsOppositePresentsDialog() throws {
        let sut = SaveFavoritesFlow(
            lineOrRoute: line,
            stop: stop,
            directions: [direction0],
            selectedDirection: 1,
            context: EditFavoritesContext.favorites,
            global: .init(objects: .init()),
            isFavorite: { _ in false },
            updateFavorites: { _ in },
            onClose: {},
            pushNavEntry: { _ in }
        )

        let exp = sut.inspection.inspect(after: 0.5) { view in
            XCTAssertNotNil(try view.find(FavoriteConfirmationDialog.self))
        }

        ViewHosting.host(view: sut.withFixedSettings([:]))

        wait(for: [exp], timeout: 2)
    }

    func testFavoritingWhenOnlyDirectionHasDisclaimer() throws {
        let sut = FavoriteConfirmationDialogContents(
            lineOrRoute: line,
            stop: stop,
            directions: [direction0],
            selectedDirection: 1,
            context: .favorites,
            favoritesToSave: [direction0: nil],
            updateLocalFavorite: { _, _ in }
        )

        try XCTAssertNotNil(sut.inspect().find(text: "Westbound service only"))
    }

    func testFavoritingWhenDropOffOnlyHasDisclaimer() throws {
        let sut = FavoriteConfirmationDialogContents(
            lineOrRoute: line,
            stop: stop,
            directions: [],
            selectedDirection: 0,
            context: .favorites,
            favoritesToSave: [direction0: nil],
            updateLocalFavorite: { _, _ in }
        )

        try XCTAssertNotNil(sut.inspect().find(text: "This stop is drop-off only"))
    }

    func testFavoritingDisplaysToast() throws {
        var updateFavoritesCalledFor: [RouteStopDirection: FavoriteSettings?] = [:]
        let onToastExp = XCTestExpectation(description: "Toast displayed with expected text")

        let route = ObjectCollectionBuilder().route { route in
            route.type = RouteType.bus
        }

        let toastVM = MockToastViewModel()
        toastVM.onShowToast = { toast in
            XCTAssertEqual("Added to Favorites", toast.message)
            onToastExp.fulfill()
        }

        let sut = SaveFavoritesFlow(
            lineOrRoute: LineOrRoute.route(route),
            stop: stop,
            directions: [direction0],
            selectedDirection: 0,
            context: EditFavoritesContext.favorites,
            global: .init(objects: .init()),
            isFavorite: { _ in false },
            updateFavorites: { updateFavoritesCalledFor = $0 },
            onClose: {},
            pushNavEntry: { _ in },
            toastVM: toastVM,
        )

        ViewHosting.host(view: sut.withFixedSettings([:]))

        wait(for: [onToastExp], timeout: 2)
        XCTAssertEqual(updateFavoritesCalledFor, [.init(route: route.id, stop: stop.id, direction: 0): .init()])
    }

    func testOpensSavePageWhenNotificationsFlagIsOnForSingleDirection() throws {
        let onCloseExp = XCTestExpectation(description: "On close called")
        let onPushNavExp = XCTestExpectation(description: "Navigation pushed")
        var pushedNav: SheetNavigationStackEntry?

        let route = ObjectCollectionBuilder().route { route in
            route.type = RouteType.bus
        }

        let sut = SaveFavoritesFlow(
            lineOrRoute: LineOrRoute.route(route),
            stop: stop,
            directions: [direction0],
            selectedDirection: 0,
            context: EditFavoritesContext.favorites,
            global: .init(objects: .init()),
            isFavorite: { _ in false },
            updateFavorites: { _ in XCTFail("Favorites should not be updated automatically") },
            onClose: { onCloseExp.fulfill() },
            pushNavEntry: { entry in
                pushedNav = entry
                onPushNavExp.fulfill()
            },
        )

        ViewHosting.host(view: sut.withFixedSettings([.notifications: true]))

        wait(for: [onCloseExp, onPushNavExp], timeout: 2)
        XCTAssertEqual(
            pushedNav,
            .saveFavorite(routeId: route.id, stopId: stop.id, selectedDirection: 0, context: .favorites)
        )
    }

    @MainActor
    func testOpensSavePageWhenNotificationsFlagIsOnForMultiDirection() throws {
        let onCloseExp = XCTestExpectation(description: "On close called")
        let onPushNavExp = XCTestExpectation(description: "Navigation pushed")
        var pushedNav: SheetNavigationStackEntry?

        let route = ObjectCollectionBuilder().route { route in
            route.type = RouteType.bus
        }

        let sut = SaveFavoritesFlow(
            lineOrRoute: line,
            stop: stop,
            directions: [direction0, direction1],
            selectedDirection: 1,
            context: EditFavoritesContext.stopDetails,
            global: .init(objects: .init()),
            isFavorite: { _ in false },
            updateFavorites: { _ in XCTFail("Favorites should not be updated automatically") },
            onClose: { onCloseExp.fulfill() },
            pushNavEntry: { entry in
                pushedNav = entry
                onPushNavExp.fulfill()
            },
        )

        ViewHosting.host(view: sut.withFixedSettings([.notifications: true]))

        wait(for: [onCloseExp, onPushNavExp], timeout: 2)
        XCTAssertEqual(
            pushedNav,
            .saveFavorite(routeId: line.id, stopId: stop.id, selectedDirection: 1, context: .stopDetails)
        )
    }
}
